File: test/test_regression.cpp
//    This file is part of ASMC, developed by Pier Francesco Palamara.
//
//    ASMC is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    ASMC is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with ASMC.  If not, see <https://www.gnu.org/licenses/>.

#define CATCH_CONFIG_MAIN
#include "catch.hpp"

#include "ASMC.hpp"
#include "FileUtils.hpp"

#include <Eigen/Core>
#include <fmt/format.h>

#include <sstream>

TEST_CASE("test ASMC regression", "[HMM_regression]")
{
  // we only needed to set doPosteriorSums to true, but because C++ does
  // not have keyword arguments we need to go through everything
  DecodingParams params(ASMC_DATA_DIR "/examples/asmc/exampleFile.n300.array",
                        ASMC_DATA_DIR "/decoding_quantities/30-100-2000_CEU.decodingQuantities.gz",
                        "",      // _outFileRoot
                        1,       // _jobs
                        1,       // _jobInd
                        "array", // _decodingModeString
                        false,   // _decodingSequence
                        true,    // _usingCSFS
                        false,   // _compress
                        false,   // _useAncestral
                        0.f,     // _skipCSFSdistance
                        false,   // _noBatches
                        true     // _doPosteriorSums
  );

  std::vector<unsigned long> indToDecodeA = {1ul, 2ul, 3ul};
  std::vector<unsigned long> indToDecodeB = {2ul, 3ul, 4ul};

  ASMC::ASMC asmc(params);
  asmc.setStorePerPairPosteriorMean();
  asmc.setStorePerPairMap();
  asmc.decodePairs(indToDecodeA, indToDecodeB);

  auto res = asmc.getRefOfResults();

  SECTION("regression test per pair posterior means")
  {

    CHECK(res.perPairPosteriorMeans.rows() == 3ll);
    CHECK(res.perPairPosteriorMeans.cols() == 6760ll);

    std::string regressionFile = ASMC_DATA_DIR "/testing/asmc/regression/regression.perPairPosteriorMeans.gz";
    FileUtils::AutoGzIfstream fin;
    fin.openOrExit(regressionFile);

    for (auto rowIdx = 0ul; rowIdx < indToDecodeA.size(); ++rowIdx) {
      std::string line;
      getline(fin, line);
      std::istringstream iss(line);
      std::vector<float> rowAsFloats = {std::istream_iterator<float>(iss), std::istream_iterator<float>()};

      CHECK(rowAsFloats.size() == 6760ul);
      for (auto colIdx = 0ul; colIdx < rowAsFloats.size(); ++colIdx) {
        CHECK(res.perPairPosteriorMeans(rowIdx, colIdx) == Approx(rowAsFloats.at(colIdx)));
      }
    }
  }

  SECTION("regression test per pair MAP")
  {
    CHECK(res.perPairMAPs.rows() == 3ll);
    CHECK(res.perPairMAPs.cols() == 6760ll);

    std::string regressionFile = ASMC_DATA_DIR "/testing/asmc/regression/regression.perPairMAP.gz";
    FileUtils::AutoGzIfstream fin;
    fin.openOrExit(regressionFile);

    for (auto rowIdx = 0ul; rowIdx < indToDecodeA.size(); ++rowIdx) {
      std::string line;
      getline(fin, line);
      std::istringstream iss(line);
      std::vector<int> rowAsInts = {std::istream_iterator<int>(iss), std::istream_iterator<int>()};

      CHECK(rowAsInts.size() == 6760ul);
      for (auto colIdx = 0ul; colIdx < rowAsInts.size(); ++colIdx) {
        CHECK(res.perPairMAPs(rowIdx, colIdx) == rowAsInts.at(colIdx));
      }
    }
  }
}

File: test/test_regression.py
import pathlib
import sys
import unittest

import numpy as np

from asmc.asmc import *

data_dir = pathlib.Path(__file__).resolve().parent.parent / 'ASMC_data'
if not data_dir.exists():
    print(f'ERROR. {data_dir} does not exist. Did you clone ASMC recursively with submodules?')
    sys.exit()


class TestASMCRegression(unittest.TestCase):

    def setUp(self):
        in_file_root = str(data_dir / 'examples' / 'asmc' / 'exampleFile.n300.array')
        decoding_quant_file = str(data_dir / 'decoding_quantities' / '30-100-2000_CEU.decodingQuantities.gz')

        params = DecodingParams(in_file_root, decoding_quant_file, do_posterior_sums=True)

        self.asmc = ASMC(params)
        self.asmc.set_store_per_pair_posterior_mean(True)
        self.asmc.set_store_per_pair_map(True)

    def test_regression(self):
        self.asmc.decode_pairs([1, 2, 3], [2, 3, 4])
        res = self.asmc.get_ref_of_results()

        self.assertEqual(res.per_pair_posterior_means.shape[0], 3)
        self.assertEqual(res.per_pair_posterior_means.shape[1], 6760)

        existing_post = np.loadtxt(
            str(data_dir / 'testing' / 'asmc' / 'regression' / 'regression.perPairPosteriorMeans.gz'))
        self.assertEqual(np.allclose(res.per_pair_posterior_means, existing_post), True)

        existing_map = np.loadtxt(
            str(data_dir / 'testing' / 'asmc' / 'regression' / 'regression.perPairMAP.gz'))
        self.assertEqual(np.allclose(res.per_pair_MAPs, existing_map), True)


class TestFastSMCRegression(unittest.TestCase):

    def setUp(self):
        # Create decoding params object with required options
        self.params = DecodingParams()
        self.params.decodingQuantFile = str(data_dir / 'decoding_quantities' / '10-20-2000_CEU.decodingQuantities.gz')
        self.params.inFileRoot = str(data_dir / 'examples' / 'fastsmc' / 'example')
        self.params.outFileRoot = '/tmp/FastSMCresults'
        self.params.decodingModeString = 'array'
        self.params.usingCSFS = True
        self.params.batchSize = 32
        self.params.recallThreshold = 3
        self.params.min_m = 1.5
        self.params.hashing = True
        self.params.FastSMC = True
        self.params.BIN_OUT = False
        self.params.outputIbdSegmentLength = True
        self.params.time = 50
        self.params.noConditionalAgeEstimates = True
        self.params.doPerPairMAP = True
        self.params.doPerPairPosteriorMean = True
        self.params.useKnownSeed = True

        assert self.params.validateParamsFastSMC()

        fast_smc = FastSMC(self.params)
        fast_smc.run()

    def test_regression(self):
        original_text = np.loadtxt(str(data_dir / 'testing' / 'fastsmc' / 'regression' / 'regression_output.ibd.gz'),
                                   usecols=(7, 8, 9, 10, 11))
        generated_text = np.loadtxt(self.params.outFileRoot + ".1.1.FastSMC.ibd.gz", usecols=(7, 8, 9, 10, 11))

        self.assertEqual(original_text.shape, generated_text.shape)
        self.assertEqual(np.allclose(original_text, generated_text), True)


class TestFastSMCRegressionWithoutHashing(unittest.TestCase):

    def setUp(self):
        # Create decoding params object with required options
        self.params = DecodingParams()
        self.params.decodingQuantFile = str(data_dir / 'decoding_quantities' / '10-20-2000_CEU.decodingQuantities.gz')
        self.params.inFileRoot = str(data_dir / 'examples' / 'fastsmc' / 'example')
        self.params.outFileRoot = '/tmp/FastSMCresults'
        self.params.decodingModeString = 'array'
        self.params.usingCSFS = True
        self.params.batchSize = 32
        self.params.recallThreshold = 3
        self.params.min_m = 1.5
        self.params.hashing = False
        self.params.FastSMC = True
        self.params.BIN_OUT = False
        self.params.outputIbdSegmentLength = True
        self.params.time = 50
        self.params.noConditionalAgeEstimates = True
        self.params.doPerPairMAP = True
        self.params.doPerPairPosteriorMean = True
        self.params.jobInd = 7
        self.params.jobs = 25
        self.params.useKnownSeed = True

        assert self.params.validateParamsFastSMC()

        fast_smc = FastSMC(self.params)
        fast_smc.run()

    def test_regression(self):
        original_text = np.loadtxt(
            str(data_dir / 'testing' / 'fastsmc' / 'regression' / 'regression_output_no_hashing.ibd.gz'),
            usecols=(7, 8, 9, 10, 11))
        generated_text = np.loadtxt(self.params.outFileRoot + ".7.25.FastSMC.ibd.gz", usecols=(7, 8, 9, 10, 11))

        self.assertEqual(original_text.shape, generated_text.shape)
        self.assertEqual(np.allclose(original_text, generated_text), True)
